routing/control_tower.go

```diff
@@ -117,20 +117,27 @@ func (p *controlTower) SettleAttempt(paymentHash lntypes.Hash,
 		return err
 	}
 
-	// Notify subscribers of success event.
-	p.notifyFinalEvent(
-		paymentHash, createSuccessResult(payment.HTLCs),
-	)
-
+	// If settling this attempt took the payment to a final state, we
+	// should notify all subscribers of success event.
+	p.notifyFinalEvent(payment)
 	return nil
 }
 
 // FailAttempt marks the given payment attempt failed.
 func (p *controlTower) FailAttempt(paymentHash lntypes.Hash,
 	attemptID uint64, failInfo *channeldb.HTLCFailInfo) error {
 
-	_, err := p.db.FailAttempt(paymentHash, attemptID, failInfo)
-	return err
+	payment, err := p.db.FailAttempt(paymentHash, attemptID, failInfo)
+	if err != nil {
+		return err
+	}
+
+	// If failing this attempt took the payment to a final state, we
+	// should notify all subscribers of the final payment status. This can
+	// happen if the payment already had a settled attempt or terminal
+	// failure recorded, and this attempt was the last one in-flight.
+	p.notifyFinalEvent(payment)
```

Comment:
Isn't this causing a duplicate event?

Reply:
Idea is that only the first event that take the payment to `Settle/failed` state will produce the notification. The second one will find the payment also in this state, but by that time the subscribers list will be empty (and cannot be filled again since we only add to it for in-flight payments)

Reply:
And there is no race condition possible there?

Reply:
I think the first one that gets the lock on the subscribers will empty the list, and at that point it is impossible to add more to the list, so I think it is okay.

If we merge #3970 first I can add some more testcases here for MPP scenarios.

Reply:
I was actually thinking about a race between subscribe and notify, because the payment update in `FailAttempt` isn't inside the subscriber lock. 

I think the bottom line of this discussion is that the concurrent behavior isn't very clear. There is a potentially a duplicate notification, but it is never broadcast because the subscriber lock makes sure the list is already empty.

Also control tower notification needs to be updated anyway to sends updates for non-final state transitions.

Isn't a single PR on top of #3970 that fixes both things a better idea?

Reply:
in order to be totally consistent i think, yes, there would need to be a lock around the db operation as well. this is preexisting tho, no?

Reply:
This is the direction that I am thinking of atm: https://github.com/lightningnetwork/lnd/compare/master...joostjager:notify-htlc-arrival-2